rustler_bigint/src/big_int.rs

```diff
@@ -0,0 +1,396 @@
+use rustler::types::atom::error;
+use rustler::{Decoder, Encoder, Env, Error, NifResult, Term};
+
+use num_bigint::Sign;
+
+/// Wrapper around num-bigint that implements [Decoder](rustler::Decoder) and [Encoder](rustler::Encoder) traits
+///
+/// ```rust
+/// use rustler_bigint::{BigInt, num_bigint};
+///
+/// // convert from and to the bigint wrapper
+/// let number = num_bigint::BigInt::from(12);
+/// let implements_decode_encode = BigInt::from(number.clone());
+/// let bigint = num_bigint::BigInt::from(implements_decode_encode);
+/// assert_eq!(number, bigint);
+/// ```
+///
+/// ## Examples
+///
+/// ```rust
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn pow(base: BigInt, exponent: u32) -> BigInt {
+///   base.pow(exponent).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::Binary;
+/// use rustler_bigint::{BigInt, num_bigint};
+///
+/// #[rustler::nif]
+/// pub fn binary_to_integer(binary: Binary) -> BigInt {
+///   num_bigint::BigInt::from_signed_bytes_be(binary.as_slice()).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::{Binary, Env, NewBinary};
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn integer_to_binary<'a>(env: Env<'a>, integer: BigInt) -> Binary<'a> {
+///   let bytes = integer.to_signed_bytes_be();
+///   let mut output = NewBinary::new(env, bytes.len());
+///   output.as_mut_slice().copy_from_slice(bytes.as_slice());
+///   output.into()
+/// }
+/// ```
+///
+#[derive(Debug, PartialEq, PartialOrd)]
+pub struct BigInt(num_bigint::BigInt);
+
+impl std::convert::From<num_bigint::BigInt> for BigInt {
+    fn from(big_int: num_bigint::BigInt) -> BigInt {
+        BigInt(big_int)
+    }
+}
+
+impl std::convert::From<BigInt> for num_bigint::BigInt {
+    fn from(big_int: BigInt) -> num_bigint::BigInt {
+        big_int.0
+    }
+}
+
+impl std::ops::Deref for BigInt {
+    type Target = num_bigint::BigInt;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for BigInt {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+fn decode_big_integer(input: &[u8]) -> NifResult<num_bigint::BigInt> {
+    if input[0] != 131 {
```

Comment:
We should also define a constant for that. It appears this is the external term format version number, so lets use that: `const EXTERNAL_TERM_FORMAT_VERSION = 131u8`;

Reply:
Also: Do we need to allow `input[0] >= 131`? I would expect that the format will be backwards compatible.

Reply:
I can't find on the internet if there ever was a different version. So maybe it is better to keep it like this and if another version comes out, re-evaluate this.

Reply:
Makes sense.